File: src/lib/utils/Writer.ts
import { Float16Array } from "@petamoriken/float16";

let GLOBAL_ENDIANNESS = true;

const encoder = new TextEncoder();

/**
 * A binary Writer for ease of use
 */
export class Writer {
  data: ArrayBuffer;
  offset: number;
  length: number;
  view: DataView;

  constructor(data: Uint8Array | Int8Array | ArrayBuffer) {
    this.data = this.#isUint8Array(data) ? data.buffer : data;
    this.view = new DataView(this.data);
    this.length = new Uint8Array(this.data).length;
    this.offset = 0;
  }

  #isUint8Array(data: Uint8Array | ArrayBuffer): data is Uint8Array {
    return (data as Uint8Array).buffer !== undefined;
  }

  private expandCapacity() {
    const newDat = new ArrayBuffer(this.length * 2);
    const uint8Arr = new Uint8Array(newDat);
    uint8Arr.set(new Uint8Array(this.data), 0);

    this.data = newDat;
    this.view = new DataView(this.data);
    this.length = new Uint8Array(this.data).length;
  }

  /**
   * Removes all trailing 0x00 bytes from the writer
   */
  trim() {
    const newDat = this.data.slice(0, this.offset);

    this.data = newDat;
    this.view = new DataView(this.data);
    this.length = new Uint8Array(this.data).length;
  }

  #writeI(
    method: keyof DataView,
    length: number
  ): (data: any, endianness?: boolean) => number {
    if (this.remaining() <= length) {
      this.expandCapacity();
    }
    return (data: any, endianness?: boolean) => {
      // @ts-ignore
      this.view[method](
        this.offset,
        data,
        endianness ? endianness : GLOBAL_ENDIANNESS
      );
      this.offset += length;
      return length;
    };
  }

  /**
   * seeks to the current offset
   * @param  {number} offset the new offset.
   * @param  {number} position the position to update from. 0 = start, 1 = current offset, 2 = end.
   */
  seek(offset: number, position: number = 0) {
    if (position == 0) {
      this.offset = Number(offset);
    } else if (position == 1) {
      this.offset = this.offset + Number(offset);
    } else if (position == 2) {
      this.offset = Number(this.data.byteLength) - Number(offset);
    } else {
      throw Error(
        `Unexpected position value. Expected 0, 1, or 2, but got ${position}.`
      );
    }
  }

  /**
   * Returns the number of bytes left in the Writer
   * @returns the number of bytes left in the Writer
   */
  remaining(): number {
    return this.length - (this.offset + 1);
  }

  /**
   * Write a byte to the current offset.
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   */
  writeByte = this.#writeI("setInt8", 1);

  /**
   * Writes the provide char to the current offset of the Writer. Returns the number of bytes written
   * @param  {string} data the char to write
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   * @returns the number of bytes written
   */
  writeChar(data: string, endianness?: boolean) {
    return this.#writeI("setInt8", 1)(data.charCodeAt(0), endianness);
  }

  /**
   * Writes the signed data to the current offset.
   * @param  {Int8Array} data the data to write
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   */
  writeSignedBytes(data: Int8Array, endianness?: boolean) {
    if (this.remaining() <= data.length) {
      this.expandCapacity();
    }
    const nDat = new Int8Array(this.data);
    nDat.set(
      (endianness ? endianness : GLOBAL_ENDIANNESS) ? data : data.reverse(),
      this.offset
    );
    this.data = nDat.buffer;
    this.view = new DataView(this.data);
    this.offset += data.length;
    return data.length;
  }

  /**
   * Writes the unsigned data to the current offset.
   * @param  {Uint8Array} data the data to write
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   */
  writeUnsignedBytes(data: Uint8Array, endianness?: boolean) {
    if (this.remaining() <= data.length) {
      this.expandCapacity();
    }
    const nDat = new Uint8Array(this.data);
    nDat.set(
      (endianness ? endianness : GLOBAL_ENDIANNESS) ? data : data.reverse(),
      this.offset
    );
    this.data = nDat.buffer;
    this.view = new DataView(this.data);
    this.offset += data.length;
    return data.length;
  }

  /**
   * Writes the provided Uint8 to the current offset
   * @param {number} data the number to write
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   * @returns the number of bytes written
   */
  writeUint8 = this.#writeI("setUint8", 1);

  /**
   * Writes the provided Uint16 to the current offset
   * @param {number} data the number to write
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   * @returns the number of bytes written
   */
  writeUint16 = this.#writeI("setUint16", 2);

  /**
   * Writes the provided Uint32 to the current offset
   * @param {number} data the number to write
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   * @returns the number of bytes written
   */
  writeUint32 = this.#writeI("setUint32", 4);

  /**
   * Writes the provided BigUint64 to the current offset
   * @param {number} data the number to write
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   * @returns the number of bytes written
   */
  writeUint64 = this.#writeI("setBigUint64", 8);

  /**
   * Writes the provided Int8 to the current offset
   * @param {number} data the number to write
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   * @returns the number of bytes written
   */
  writeInt8 = this.#writeI("setInt8", 1);

  /**
   * Writes the provided Int16 to the current offset
   * @param {number} data the number to write
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   * @returns the number of bytes written
   */
  writeInt16 = this.#writeI("setInt16", 2);

  /**
   * Writes the provided Int32 to the current offset
   * @param {number} data the number to write
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   * @returns the number of bytes written
   */
  writeInt32 = this.#writeI("setInt32", 4);

  /**
   * Writes the provided BigInt64 to the current offset
   * @param {number} data the number to write
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   * @returns the number of bytes written
   */
  writeInt64 = this.#writeI("setBigInt64", 8);

  /**
   * Writes the provided Float16 to the current offset
   * @param {number} data the number to write
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   * @returns the number of bytes written
   */
  writeFloat16(data: number, endianness: boolean = true) {
    if (this.remaining() <= 2) {
      this.expandCapacity();
    }
    const res = new Float16Array([data]);

    const nDat = new Float16Array(this.data);
    nDat.set(
      (endianness ? endianness : GLOBAL_ENDIANNESS) ? res : res.reverse(),
      this.offset
    );
    this.data = nDat.buffer;
    this.view = new DataView(this.data);

    this.offset += 2;
    return 1;
  }

  /**
   * Writes the provided Float32 to the current offset
   * @param {number} data the number to write
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   * @returns the number of bytes written
   */
  writeFloat32 = this.#writeI("setFloat32", 4);

  /**
   * Writes the provided Float64 to the current offset
   * @param {number} data the number to write
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   * @returns the number of bytes written
   */
  writeFloat64 = this.#writeI("setFloat64", 8);

  /**
   * Writes the provided string to the current offset, prefixed by the length as an Int32
   * @param {string} str the string to write
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   * @returns the number of bytes written
   */
  writeLenPrefixString(str: string, endianness: boolean = true): number {
    const length = str.length;
    if (this.remaining() <= length) {
      this.expandCapacity();
    }
    this.writeUint32(length, endianness);

    const strBytes = encoder.encode(str);
    this.writeUnsignedBytes(strBytes, endianness);

    return 4 + strBytes.length;
  }

  /**
   * Writes the provided string to the current offset.
   * @param {string} str the string to write
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   * @returns the number of bytes written
   */
  writeString(str: string, endianness: boolean = true): number {
    const length = str.length;
    if (this.remaining() <= length) {
      this.expandCapacity();
    }

    const strBytes = encoder.encode(str);
    this.writeUnsignedBytes(strBytes, endianness);

    return strBytes.length;
  }

  /**
   * Writes the provided string to the current offset, followed by a 0x00 byte
   * @param {string} str the string to write
   * @param  {boolean} endianness whether or not to use littleEdian. Default is true.
   * @returns the number of bytes written
   */
  write00PaddedString(str: string, endianness: boolean = true): number {
    const strBytes = encoder.encode(str);
    if (this.remaining() <= strBytes.length) {
      this.expandCapacity();
    }
    this.writeUnsignedBytes(strBytes, endianness);

    this.writeUint8(0x00, endianness);

    return strBytes.length + 1;
  }
}

File: src-tauri/src/writer.rs
trait HasByteConvert {
  fn to_le_bytes(bytes: &mut Vec<u8>, data: Self, offset: usize) -> u8;
  fn to_be_bytes(bytes: &mut Vec<u8>, data: Self, offset: usize) -> u8;
}

impl HasByteConvert for u8 {
  fn to_le_bytes(bytes: &mut Vec<u8>, data: u8, offset: usize) -> u8 {
    bytes[offset] = data;
    return 1;
  }
  fn to_be_bytes(bytes: &mut Vec<u8>, data: u8, offset: usize) -> u8 {
    bytes[offset] = data;
    return 1;
  }
}

impl HasByteConvert for u16 {
  fn to_le_bytes(bytes: &mut Vec<u8>, data: u16, offset: usize) -> u8 {
    let u8_arr = data.to_le_bytes();
    bytes[offset] = u8_arr[0];
    bytes[offset+1] = u8_arr[1];
    return 2;
  }
  fn to_be_bytes(bytes: &mut Vec<u8>, data: u16, offset: usize) -> u8 {
    let u8_arr = data.to_be_bytes();
    bytes[offset] = u8_arr[0];
    bytes[offset+1] = u8_arr[1];
    return 2;
  }
}

impl HasByteConvert for u32 {
  fn to_le_bytes(bytes: &mut Vec<u8>, data: u32, offset: usize) -> u8 {
    let u8_arr = data.to_le_bytes();
    bytes[offset] = u8_arr[0];
    bytes[offset+1] = u8_arr[1];
    bytes[offset+2] = u8_arr[2];
    bytes[offset+3] = u8_arr[3];
    return 4;
  }
  fn to_be_bytes(bytes: &mut Vec<u8>, data: u32, offset: usize) -> u8 {
    let u8_arr = data.to_be_bytes();
    bytes[offset] = u8_arr[0];
    bytes[offset+1] = u8_arr[1];
    bytes[offset+2] = u8_arr[2];
    bytes[offset+3] = u8_arr[3];
    return 4;
  }
}

impl HasByteConvert for u64 {
  fn to_le_bytes(bytes: &mut Vec<u8>, data: u64, offset: usize) -> u8 {
    let u8_arr = data.to_le_bytes();
    bytes[offset] = u8_arr[0];
    bytes[offset+1] = u8_arr[1];
    bytes[offset+2] = u8_arr[2];
    bytes[offset+3] = u8_arr[3];
    bytes[offset+4] = u8_arr[4];
    bytes[offset+5] = u8_arr[5];
    bytes[offset+6] = u8_arr[6];
    bytes[offset+7] = u8_arr[7];
    return 8;
  }
  fn to_be_bytes(bytes: &mut Vec<u8>, data: u64, offset: usize) -> u8 {
    let u8_arr = data.to_be_bytes();
    bytes[offset] = u8_arr[0];
    bytes[offset+1] = u8_arr[1];
    bytes[offset+2] = u8_arr[2];
    bytes[offset+3] = u8_arr[3];
    bytes[offset+4] = u8_arr[4];
    bytes[offset+5] = u8_arr[5];
    bytes[offset+6] = u8_arr[6];
    bytes[offset+7] = u8_arr[7];
    return 8;
  }
}

impl HasByteConvert for i8 {
  fn to_le_bytes(bytes: &mut Vec<u8>, data: i8, offset: usize) -> u8 {
    bytes[offset] = data.to_le_bytes()[0];
    return 1;
  }
  fn to_be_bytes(bytes: &mut Vec<u8>, data: i8, offset: usize) -> u8 {
    bytes[offset] = data.to_be_bytes()[0];
    return 1;
  }
}

impl HasByteConvert for i16 {
  fn to_le_bytes(bytes: &mut Vec<u8>, data: i16, offset: usize) -> u8 {
    let u8_arr = data.to_le_bytes();
    bytes[offset] = u8_arr[0];
    bytes[offset+1] = u8_arr[1];
    return 2;
  }
  fn to_be_bytes(bytes: &mut Vec<u8>, data: i16, offset: usize) -> u8 {
    let u8_arr = data.to_be_bytes();
    bytes[offset] = u8_arr[0];
    bytes[offset+1] = u8_arr[1];
    return 2;
  }
}

impl HasByteConvert for i32 {
  fn to_le_bytes(bytes: &mut Vec<u8>, data: i32, offset: usize) -> u8 {
    let u8_arr = data.to_le_bytes();
    bytes[offset] = u8_arr[0];
    bytes[offset+1] = u8_arr[1];
    bytes[offset+2] = u8_arr[2];
    bytes[offset+3] = u8_arr[3];
    return 4;
  }
  fn to_be_bytes(bytes: &mut Vec<u8>, data: i32, offset: usize) -> u8 {
    let u8_arr = data.to_be_bytes();
    bytes[offset] = u8_arr[0];
    bytes[offset+1] = u8_arr[1];
    bytes[offset+2] = u8_arr[2];
    bytes[offset+3] = u8_arr[3];
    return 4;
  }
}

impl HasByteConvert for i64 {
  fn to_le_bytes(bytes: &mut Vec<u8>, data: i64, offset: usize) -> u8 {
    let u8_arr = data.to_le_bytes();
    bytes[offset] = u8_arr[0];
    bytes[offset+1] = u8_arr[1];
    bytes[offset+2] = u8_arr[2];
    bytes[offset+3] = u8_arr[3];
    bytes[offset+4] = u8_arr[4];
    bytes[offset+5] = u8_arr[5];
    bytes[offset+6] = u8_arr[6];
    bytes[offset+7] = u8_arr[7];
    return 8;
  }
  fn to_be_bytes(bytes: &mut Vec<u8>, data: i64, offset: usize) -> u8 {
    let u8_arr = data.to_be_bytes();
    bytes[offset] = u8_arr[0];
    bytes[offset+1] = u8_arr[1];
    bytes[offset+2] = u8_arr[2];
    bytes[offset+3] = u8_arr[3];
    bytes[offset+4] = u8_arr[4];
    bytes[offset+5] = u8_arr[5];
    bytes[offset+6] = u8_arr[6];
    bytes[offset+7] = u8_arr[7];
    return 8;
  }
}

impl HasByteConvert for f32 {
  fn to_le_bytes(bytes: &mut Vec<u8>, data: f32, offset: usize) -> u8 {
    let u8_arr = data.to_le_bytes();
    bytes[offset] = u8_arr[0];
    bytes[offset+1] = u8_arr[1];
    bytes[offset+2] = u8_arr[2];
    bytes[offset+3] = u8_arr[3];
    return 4;
  }
  fn to_be_bytes(bytes: &mut Vec<u8>, data: f32, offset: usize) -> u8 {
    let u8_arr = data.to_be_bytes();
    bytes[offset] = u8_arr[0];
    bytes[offset+1] = u8_arr[1];
    bytes[offset+2] = u8_arr[2];
    bytes[offset+3] = u8_arr[3];
    return 4;
  }
}

impl HasByteConvert for f64 {
  fn to_le_bytes(bytes: &mut Vec<u8>, data: f64, offset: usize) -> u8 {
    let u8_arr = data.to_le_bytes();
    bytes[offset] = u8_arr[0];
    bytes[offset+1] = u8_arr[1];
    bytes[offset+2] = u8_arr[2];
    bytes[offset+3] = u8_arr[3];
    bytes[offset+4] = u8_arr[4];
    bytes[offset+5] = u8_arr[5];
    bytes[offset+6] = u8_arr[6];
    bytes[offset+7] = u8_arr[7];
    return 8;
  }
  fn to_be_bytes(bytes: &mut Vec<u8>, data: f64, offset: usize) -> u8 {
    let u8_arr = data.to_be_bytes();
    bytes[offset] = u8_arr[0];
    bytes[offset+1] = u8_arr[1];
    bytes[offset+2] = u8_arr[2];
    bytes[offset+3] = u8_arr[3];
    bytes[offset+4] = u8_arr[4];
    bytes[offset+5] = u8_arr[5];
    bytes[offset+6] = u8_arr[6];
    bytes[offset+7] = u8_arr[7];
    return 8;
  }
}

pub struct Writer<'a> {
  data: &'a mut Vec<u8>,
  offset: usize,
}

#[allow(dead_code)]
impl Writer<'_> {
  pub fn get_data(&self) -> &[u8] { return &self.data[..]; }
  pub fn get_offset(&self) -> usize { return self.offset; }

  pub fn new(buf: &mut Vec<u8>) -> Writer { return Writer { data: buf, offset: 0 }; }

  pub fn seek(&mut self, offset: usize, position: u8) {
    if position == 0 {
      self.offset = offset;
    } else if position == 1 {
      self.offset += offset;
    } else {
      self.offset = (self.data.len() as usize) - offset;
    }
  }
  
  fn expand_capacity(&mut self) {
    self.data.resize(self.data.len() * 2, 0);
  }

  fn write_i<T: HasByteConvert>(&mut self, data: T, length: u8, endianness: bool) -> u8 {
    if self.remaining() <= length.into() {
      self.expand_capacity()
    }

    if endianness {
      return T::to_le_bytes(self.data, data, self.offset);
    } else {
      return T::to_be_bytes(self.data, data, self.offset);
    }
  }

  pub fn trim(&mut self) {
    self.data.truncate(self.offset);
  }

  pub fn remaining(&mut self) -> usize {
    return self.data.len() - (self.offset + 1);
  }
  
  pub fn write_uint8(&mut self, data: u8, endianness: bool) -> u8 {
    let res = self.write_i::<u8>(data, 1, endianness);
    self.offset += 1;
    return res;
  }
  pub fn write_uint16(&mut self, data: u16, endianness: bool) -> u8 {
    let res = self.write_i::<u16>(data, 2, endianness);
    self.offset += 2;
    return res;
  }
  pub fn write_uint32(&mut self, data: u32, endianness: bool) -> u8 {
    let res = self.write_i::<u32>(data, 4, endianness);
    self.offset += 4;
    return res;
  }
  pub fn write_uint64(&mut self, data: u64, endianness: bool) -> u8 {
    let res = self.write_i::<u64>(data, 8, endianness);
    self.offset += 8;
    return res;
  }
  
  pub fn write_int8(&mut self, data: i8, endianness: bool) -> u8 {
    let res = self.write_i::<i8>(data, 1, endianness);
    self.offset += 1;
    return res;
  }
  pub fn write_int16(&mut self, data: i16, endianness: bool) -> u8 {
    let res = self.write_i::<i16>(data, 2, endianness);
    self.offset += 2;
    return res;
  }
  pub fn write_int32(&mut self, data: i32, endianness: bool) -> u8 {
    let res = self.write_i::<i32>(data, 4, endianness);
    self.offset += 4;
    return res;
  }
  pub fn write_int64(&mut self, data: i64, endianness: bool) -> u8 {
    let res = self.write_i::<i64>(data, 8, endianness);
    self.offset += 8;
    return res;
  }
  
  pub fn write_float32(&mut self, data: f32, endianness: bool) -> u8 {
    let res = self.write_i::<f32>(data, 4, endianness);
    self.offset += 4;
    return res;
  }
  pub fn write_float64(&mut self, data: f64, endianness: bool) -> u8 {
    let res = self.write_i::<f64>(data, 8, endianness);
    self.offset += 8;
    return res;
  }

  pub fn write_string(&mut self, data: String, prefix_with_length: bool, endianness: bool) -> u32 {
    let str_bytes = data.into_bytes();
    let str_bytes_len = str_bytes.len();
    let mut length_to_return = str_bytes_len;

    if prefix_with_length {
      self.write_uint32(str_bytes_len as u32, endianness);
      length_to_return += 4;
    }
    
    if self.remaining() <= self.data.len() {
      self.expand_capacity()
    }
    
    let mut i: usize = 0;
    while i < str_bytes_len {
      self.data[self.offset + i] = str_bytes[i];
      i += 1;
    }

    self.data[self.offset + str_bytes_len] = 0x00;

    self.offset += str_bytes_len;

    return length_to_return as u32;
  }
}